File: py/aviationweatherdata_sdk.py
# AviationweatherData SDK

from utility.voxgig_struct import voxgig_struct as vs
from core.utility_type import AviationweatherDataUtility
from core.spec import AviationweatherDataSpec
from core import helpers

# Load utility registration (populates Utility._registrar)
from utility import register

# Load features
from feature.base_feature import AviationweatherDataBaseFeature
from features import _make_feature


class AviationweatherDataSDK:

    def __init__(self, options=None):
        self.mode = "live"
        self.features = []
        self.options = None

        utility = AviationweatherDataUtility()
        self._utility = utility

        from config import make_config
        config = make_config()

        self._rootctx = utility.make_context({
            "client": self,
            "utility": utility,
            "config": config,
            "options": options if options is not None else {},
            "shared": {},
        }, None)

        self.options = utility.make_options(self._rootctx)

        if vs.getpath(self.options, "feature.test.active") is True:
            self.mode = "test"

        self._rootctx.options = self.options

        # Add features in the resolved order (make_options puts an explicit
        # list order first, else defaults to test-first). Ordering matters: the
        # `test` feature installs the base mock transport and the transport
        # features (retry/cache/netsim/proxy/ratelimit) wrap whatever is
        # current, so `test` must be added before them to sit at the base.
        feature_opts = helpers.to_map(vs.getprop(self.options, "feature"))
        if feature_opts is not None:
            featureorder = vs.getpath(self.options, "__derived__.featureorder")
            if isinstance(featureorder, list):
                for fname in featureorder:
                    fopts = helpers.to_map(feature_opts.get(fname))
                    if fopts is not None and fopts.get("active") is True:
                        utility.feature_add(self._rootctx, _make_feature(fname))

        # Add extension features.
        extend = vs.getprop(self.options, "extend")
        if isinstance(extend, list):
            for f in extend:
                if isinstance(f, dict) or (hasattr(f, "get_name") and callable(f.get_name)):
                    utility.feature_add(self._rootctx, f)

        # Initialize features.
        for f in self.features:
            utility.feature_init(self._rootctx, f)

        utility.feature_hook(self._rootctx, "PostConstruct")

        # #BuildFeatures

    def options_map(self):
        out = vs.clone(self.options)
        if isinstance(out, dict):
            return out
        return {}

    def get_utility(self):
        return AviationweatherDataUtility.copy(self._utility)

    def get_root_ctx(self):
        return self._rootctx

    def prepare(self, fetchargs=None):
        utility = self._utility

        if fetchargs is None:
            fetchargs = {}

        ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl"))
        if ctrl is None:
            ctrl = {}

        ctx = utility.make_context({
            "opname": "prepare",
            "ctrl": ctrl,
        }, self._rootctx)

        options = self.options

        path = vs.getprop(fetchargs, "path") or ""
        if not isinstance(path, str):
            path = ""

        method = vs.getprop(fetchargs, "method") or "GET"
        if not isinstance(method, str):
            method = "GET"

        params = helpers.to_map(vs.getprop(fetchargs, "params"))
        if params is None:
            params = {}
        query = helpers.to_map(vs.getprop(fetchargs, "query"))
        if query is None:
            query = {}

        headers = utility.prepare_headers(ctx)

        base = vs.getprop(options, "base") or ""
        if not isinstance(base, str):
            base = ""
        prefix = vs.getprop(options, "prefix") or ""
        if not isinstance(prefix, str):
            prefix = ""
        suffix = vs.getprop(options, "suffix") or ""
        if not isinstance(suffix, str):
            suffix = ""

        ctx.spec = AviationweatherDataSpec({
            "base": base,
            "prefix": prefix,
            "suffix": suffix,
            "path": path,
            "method": method,
            "params": params,
            "query": query,
            "headers": headers,
            "body": vs.getprop(fetchargs, "body"),
            "step": "start",
        })

        # Merge user-provided headers.
        uh = vs.getprop(fetchargs, "headers")
        if isinstance(uh, dict):
            for k, v in uh.items():
                ctx.spec.headers[k] = v

        _, err = utility.prepare_auth(ctx)
        if err is not None:
            raise err

        fetchdef, err = utility.make_fetch_def(ctx)
        if err is not None:
            raise err

        return fetchdef

    def direct(self, fetchargs=None):
        utility = self._utility

        try:
            fetchdef = self.prepare(fetchargs)
        except Exception as err:
            # direct() is the raw-HTTP escape hatch: it never raises, it
            # returns a result object callers branch on via result["ok"].
            return {"ok": False, "err": err}

        if fetchargs is None:
            fetchargs = {}
        ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl"))
        if ctrl is None:
            ctrl = {}

        ctx = utility.make_context({
            "opname": "direct",
            "ctrl": ctrl,
        }, self._rootctx)

        url = fetchdef.get("url", "")
        fetched, fetch_err = utility.fetcher(ctx, url, fetchdef)

        if fetch_err is not None:
            return {"ok": False, "err": fetch_err}

        if fetched is None:
            return {
                "ok": False,
                "err": ctx.make_error("direct_no_response", "response: undefined"),
            }

        if isinstance(fetched, dict):
            status = helpers.to_int(vs.getprop(fetched, "status"))
            headers = vs.getprop(fetched, "headers") or {}

            # No-body responses (204, 304) and explicit zero content-length
            # must skip JSON parsing — calling json() on an empty body raises.
            content_length = None
            if isinstance(headers, dict):
                content_length = headers.get("content-length")
            no_body = status in (204, 304) or str(content_length) == "0"

            json_data = None
            if not no_body:
                jf = vs.getprop(fetched, "json")
                if callable(jf):
                    try:
                        json_data = jf()
                    except Exception:
                        # Non-JSON body (e.g. text/plain, text/html). Surface
                        # status + headers but leave data as None.
                        json_data = None

            return {
                "ok": status >= 200 and status < 300,
                "status": status,
                "headers": headers,
                "data": json_data,
            }

        return {
            "ok": False,
            "err": ctx.make_error("direct_invalid", "invalid response type"),
        }


    def AirSigmet(self, data=None) -> "AirSigmetEntity":
        """Entity factory: client.AirSigmet().list() / client.AirSigmet().load({"id": ...})."""
        from entity.air_sigmet_entity import AirSigmetEntity
        return AirSigmetEntity(self, data)


    def Airport(self, data=None) -> "AirportEntity":
        """Entity factory: client.Airport().list() / client.Airport().load({"id": ...})."""
        from entity.airport_entity import AirportEntity
        return AirportEntity(self, data)


    def Cache(self, data=None) -> "CacheEntity":
        """Entity factory: client.Cache().list() / client.Cache().load({"id": ...})."""
        from entity.cache_entity import CacheEntity
        return CacheEntity(self, data)


    def Cwa(self, data=None) -> "CwaEntity":
        """Entity factory: client.Cwa().list() / client.Cwa().load({"id": ...})."""
        from entity.cwa_entity import CwaEntity
        return CwaEntity(self, data)


    def GAirmet(self, data=None) -> "GAirmetEntity":
        """Entity factory: client.GAirmet().list() / client.GAirmet().load({"id": ...})."""
        from entity.g_airmet_entity import GAirmetEntity
        return GAirmetEntity(self, data)


    def Metar(self, data=None) -> "MetarEntity":
        """Entity factory: client.Metar().list() / client.Metar().load({"id": ...})."""
        from entity.metar_entity import MetarEntity
        return MetarEntity(self, data)


    def Pirep(self, data=None) -> "PirepEntity":
        """Entity factory: client.Pirep().list() / client.Pirep().load({"id": ...})."""
        from entity.pirep_entity import PirepEntity
        return PirepEntity(self, data)


    def StationInfo(self, data=None) -> "StationInfoEntity":
        """Entity factory: client.StationInfo().list() / client.StationInfo().load({"id": ...})."""
        from entity.station_info_entity import StationInfoEntity
        return StationInfoEntity(self, data)


    def Taf(self, data=None) -> "TafEntity":
        """Entity factory: client.Taf().list() / client.Taf().load({"id": ...})."""
        from entity.taf_entity import TafEntity
        return TafEntity(self, data)


    def Tcf(self, data=None) -> "TcfEntity":
        """Entity factory: client.Tcf().list() / client.Tcf().load({"id": ...})."""
        from entity.tcf_entity import TcfEntity
        return TcfEntity(self, data)



    @classmethod
    def test(cls, testopts=None, sdkopts=None) -> "AviationweatherDataSDK":
        if sdkopts is None:
            sdkopts = {}
        sdkopts = vs.clone(sdkopts)
        if not isinstance(sdkopts, dict):
            sdkopts = {}

        if testopts is None:
            testopts = {}
        testopts = vs.clone(testopts)
        if not isinstance(testopts, dict):
            testopts = {}
        testopts["active"] = True

        vs.setpath(sdkopts, "feature.test", testopts)

        sdk = cls(sdkopts)
        sdk.mode = "test"

        return sdk


from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from entity.air_sigmet_entity import AirSigmetEntity
    from entity.airport_entity import AirportEntity
    from entity.cache_entity import CacheEntity
    from entity.cwa_entity import CwaEntity
    from entity.g_airmet_entity import GAirmetEntity
    from entity.metar_entity import MetarEntity
    from entity.pirep_entity import PirepEntity
    from entity.station_info_entity import StationInfoEntity
    from entity.taf_entity import TafEntity
    from entity.tcf_entity import TcfEntity

File: rb/AviationweatherData_sdk.rb
# AviationweatherData SDK

require_relative 'utility/struct/voxgig_struct'
require_relative 'core/utility_type'
require_relative 'core/spec'
require_relative 'core/helpers'

# Load utility registration
require_relative 'utility/register'

# Load config and features
require_relative 'config'
require_relative 'feature/base_feature'
require_relative 'features'

# Load typed models (Struct value objects).
require_relative 'AviationweatherData_types'


class AviationweatherDataSDK
  attr_accessor :mode, :features, :options

  def initialize(options = {})
    @mode = "live"
    @features = []
    @options = nil

    utility = AviationweatherDataUtility.new
    @_utility = utility

    config = AviationweatherDataConfig.make_config

    @_rootctx = utility.make_context.call({
      "client" => self,
      "utility" => utility,
      "config" => config,
      "options" => options || {},
      "shared" => {},
    }, nil)

    @options = utility.make_options.call(@_rootctx)

    if VoxgigStruct.getpath(@options, "feature.test.active") == true
      @mode = "test"
    end

    @_rootctx.options = @options

    # Add features in the resolved order (make_options puts an explicit array
    # order first, else defaults to test-first). Ordering matters: the `test`
    # feature installs the base mock transport and the transport features
    # (retry/cache/netsim/proxy/ratelimit) wrap whatever is current, so `test`
    # must be added before them to sit at the base of the chain.
    feature_opts = AviationweatherDataHelpers.to_map(VoxgigStruct.getprop(@options, "feature"))
    if feature_opts
      featureorder = VoxgigStruct.getpath(@options, "__derived__.featureorder")
      if featureorder.is_a?(Array)
        featureorder.each do |fname|
          fopts = AviationweatherDataHelpers.to_map(feature_opts[fname])
          if fopts && fopts["active"] == true
            utility.feature_add.call(@_rootctx, AviationweatherDataFeatures.make_feature(fname))
          end
        end
      end
    end

    # Add extension features.
    extend_val = VoxgigStruct.getprop(@options, "extend")
    if extend_val.is_a?(Array)
      extend_val.each do |f|
        if f.respond_to?(:get_name)
          utility.feature_add.call(@_rootctx, f)
        end
      end
    end

    # Initialize features.
    @features.each do |f|
      utility.feature_init.call(@_rootctx, f)
    end

    utility.feature_hook.call(@_rootctx, "PostConstruct")
  end

  def options_map
    out = VoxgigStruct.clone(@options)
    out.is_a?(Hash) ? out : {}
  end

  def get_utility
    AviationweatherDataUtility.copy(@_utility)
  end

  def get_root_ctx
    @_rootctx
  end

  def prepare(fetchargs = {})
    utility = @_utility
    fetchargs ||= {}

    ctrl = AviationweatherDataHelpers.to_map(VoxgigStruct.getprop(fetchargs, "ctrl")) || {}

    ctx = utility.make_context.call({
      "opname" => "prepare",
      "ctrl" => ctrl,
    }, @_rootctx)

    opts = @options
    path = VoxgigStruct.getprop(fetchargs, "path") || ""
    path = "" unless path.is_a?(String)
    method_val = VoxgigStruct.getprop(fetchargs, "method") || "GET"
    method_val = "GET" unless method_val.is_a?(String)
    params = AviationweatherDataHelpers.to_map(VoxgigStruct.getprop(fetchargs, "params")) || {}
    query = AviationweatherDataHelpers.to_map(VoxgigStruct.getprop(fetchargs, "query")) || {}
    headers = utility.prepare_headers.call(ctx)

    base = VoxgigStruct.getprop(opts, "base") || ""
    base = "" unless base.is_a?(String)
    prefix = VoxgigStruct.getprop(opts, "prefix") || ""
    prefix = "" unless prefix.is_a?(String)
    suffix = VoxgigStruct.getprop(opts, "suffix") || ""
    suffix = "" unless suffix.is_a?(String)

    ctx.spec = AviationweatherDataSpec.new({
      "base" => base, "prefix" => prefix, "suffix" => suffix,
      "path" => path, "method" => method_val,
      "params" => params, "query" => query, "headers" => headers,
      "body" => VoxgigStruct.getprop(fetchargs, "body"),
      "step" => "start",
    })

    # Merge user-provided headers.
    uh = VoxgigStruct.getprop(fetchargs, "headers")
    if uh.is_a?(Hash)
      uh.each { |k, v| ctx.spec.headers[k] = v }
    end

    _, err = utility.prepare_auth.call(ctx)
    raise err if err

    # make_fetch_def returns a (fetchdef, err) tuple; destructure it and
    # return just the fetchdef Hash (raising on error) so callers — including
    # direct(), which indexes fetchdef["url"] — receive a Hash, mirroring the
    # ts/py prepare().
    fetchdef, fd_err = utility.make_fetch_def.call(ctx)
    raise fd_err if fd_err

    fetchdef
  end

  def direct(fetchargs = {})
    utility = @_utility

    # direct() is the raw-HTTP escape hatch: it always returns a result hash
    # ({ "ok" => ..., ... }) and never raises. prepare() raises on error, so
    # trap that and surface it in the hash.
    begin
      fetchdef = prepare(fetchargs)
    rescue AviationweatherDataError => err
      return { "ok" => false, "err" => err }
    end

    fetchargs ||= {}
    ctrl = AviationweatherDataHelpers.to_map(VoxgigStruct.getprop(fetchargs, "ctrl")) || {}

    ctx = utility.make_context.call({
      "opname" => "direct",
      "ctrl" => ctrl,
    }, @_rootctx)

    url = fetchdef["url"] || ""
    fetched, fetch_err = utility.fetcher.call(ctx, url, fetchdef)

    return { "ok" => false, "err" => fetch_err } if fetch_err

    if fetched.nil?
      return {
        "ok" => false,
        "err" => ctx.make_error("direct_no_response", "response: undefined"),
      }
    end

    if fetched.is_a?(Hash)
      status = AviationweatherDataHelpers.to_int(VoxgigStruct.getprop(fetched, "status"))
      headers = VoxgigStruct.getprop(fetched, "headers") || {}

      # No-body responses (204, 304) and explicit zero content-length must
      # skip JSON parsing — calling json() on an empty body errors.
      content_length = headers.is_a?(Hash) ? headers["content-length"] : nil
      no_body = status == 204 || status == 304 || content_length.to_s == "0"

      json_data = nil
      unless no_body
        jf = VoxgigStruct.getprop(fetched, "json")
        if jf.is_a?(Proc)
          begin
            json_data = jf.call
          rescue StandardError
            # Non-JSON body — leave data nil, keep status/headers.
            json_data = nil
          end
        end
      end

      return {
        "ok" => status >= 200 && status < 300,
        "status" => status,
        "headers" => headers,
        "data" => json_data,
      }
    end

    return {
      "ok" => false,
      "err" => ctx.make_error("direct_invalid", "invalid response type"),
    }
  end


  # Canonical facade: client.AirSigmet.list / client.AirSigmet.load({ "id" => ... })
  def AirSigmet(data = nil)
    require_relative 'entity/air_sigmet_entity'
    AirSigmetEntity.new(self, data)
  end


  # Canonical facade: client.Airport.list / client.Airport.load({ "id" => ... })
  def Airport(data = nil)
    require_relative 'entity/airport_entity'
    AirportEntity.new(self, data)
  end


  # Canonical facade: client.Cache.list / client.Cache.load({ "id" => ... })
  def Cache(data = nil)
    require_relative 'entity/cache_entity'
    CacheEntity.new(self, data)
  end


  # Canonical facade: client.Cwa.list / client.Cwa.load({ "id" => ... })
  def Cwa(data = nil)
    require_relative 'entity/cwa_entity'
    CwaEntity.new(self, data)
  end


  # Canonical facade: client.GAirmet.list / client.GAirmet.load({ "id" => ... })
  def GAirmet(data = nil)
    require_relative 'entity/g_airmet_entity'
    GAirmetEntity.new(self, data)
  end


  # Canonical facade: client.Metar.list / client.Metar.load({ "id" => ... })
  def Metar(data = nil)
    require_relative 'entity/metar_entity'
    MetarEntity.new(self, data)
  end


  # Canonical facade: client.Pirep.list / client.Pirep.load({ "id" => ... })
  def Pirep(data = nil)
    require_relative 'entity/pirep_entity'
    PirepEntity.new(self, data)
  end


  # Canonical facade: client.StationInfo.list / client.StationInfo.load({ "id" => ... })
  def StationInfo(data = nil)
    require_relative 'entity/station_info_entity'
    StationInfoEntity.new(self, data)
  end


  # Canonical facade: client.Taf.list / client.Taf.load({ "id" => ... })
  def Taf(data = nil)
    require_relative 'entity/taf_entity'
    TafEntity.new(self, data)
  end


  # Canonical facade: client.Tcf.list / client.Tcf.load({ "id" => ... })
  def Tcf(data = nil)
    require_relative 'entity/tcf_entity'
    TcfEntity.new(self, data)
  end



  def self.test(testopts = nil, sdkopts = nil)
    sdkopts = sdkopts || {}
    sdkopts = VoxgigStruct.clone(sdkopts)
    sdkopts = {} unless sdkopts.is_a?(Hash)

    testopts = testopts || {}
    testopts = VoxgigStruct.clone(testopts)
    testopts = {} unless testopts.is_a?(Hash)
    testopts["active"] = true

    VoxgigStruct.setpath(sdkopts, "feature.test", testopts)

    sdk = AviationweatherDataSDK.new(sdkopts)
    sdk.mode = "test"
    sdk
  end
end
